crypto/params_template.c

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+#include "internal/cryptlib.h"
+#include "internal/params_template.h"
+
+typedef union align_un {
+    OSSL_UNION_ALIGN;
+} OSSL_PARAMTMPL_BLOCK;
+
+#define ALIGN_SIZE  sizeof(OSSL_PARAMTMPL_BLOCK)
+
+static size_t bytes_to_blocks(size_t bytes)
+{
+    return (bytes + ALIGN_SIZE - 1) / ALIGN_SIZE;
+}
+
+static OSSL_PARAMDEF *param_push(OSSL_PARAMTMPL *tmpl, const char *key,
+                                 size_t size, size_t alloc, int type)
+{
+    OSSL_PARAMDEF *pd;
+
+    if (tmpl->curr >= OSSL_PARAMTMPL_MAX)
+        return NULL;
+    pd = tmpl->params + tmpl->curr++;
+    memset(pd, 0, sizeof(*pd));
+    pd->key = key;
+    pd->type = type;
+    pd->size = size;
+    tmpl->total_blocks += pd->alloc_blocks = bytes_to_blocks(size);
+    return pd;
+}
+
+static int param_push_num(OSSL_PARAMTMPL *tmpl, const char *key,
+                          void *num, size_t size, int type)
+{
+    OSSL_PARAMDEF *pd = param_push(tmpl, key, size, size, type);
+
+    if (pd == NULL || size > sizeof(pd->num))
+        return 0;
+    memcpy(&pd->num, num, size);
+    return 1;
+}
+
+void ossl_paramtmpl_init(OSSL_PARAMTMPL *tmpl)
+{
+    memset(tmpl, 0, sizeof(*tmpl));
+    tmpl->ispublic = 1;
+}
+
+void ossl_paramtmpl_set_private(OSSL_PARAMTMPL *tmpl)
+{
+    tmpl->ispublic = 0;
+}
```

Comment:
I wonder why this function is needed, and why it has to be called early. Since it only affects `ossl_paramtmpl_to_param`, it could just as well be an argument to that function.

But, more on that below...

Reply:
It doesn't need to be called early, it just needs to be called if you want everything securely allocated.

It could become a parameter to the conversion function which could also check for private BN's while calculating the size.